topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,620 @@
+
+	if node == nil {
+		graphMetaData = p.getMetaData(blockdev, childCount, parentWWN)
+		var err error
+		if node, err = p.Ctx.Graph.NewNode(graph.GenID(), graphMetaData); err != nil {
+			p.Ctx.Logger.Error(err)
+			return nil
+		}
+
+		// If there is a WWN - check to see if there are other paths to the same block device.
+		if blockdev.WWN != "" {
+			multiPathNodes := p.Ctx.Graph.GetNodes(graph.Metadata{"BlockID": blockdev.getID()})
+
+			for _, multiPathNode := range multiPathNodes {
+				topology.AddLink(p.Ctx.Graph, multiPathNode, node, "multipath", nil)
+			}
+		}
+
+		topology.AddOwnershipLink(p.Ctx.Graph, groupNode, node, nil)
+
+		// Link physical disks and DVD/rom devices to the blockdevGroupName
+		if blockdev.Type == "disk" || blockdev.Type == "rom" {
+			linkMetadata := graph.Metadata{
+				"Type":      "blockdev",
+				"Bandwidth": 120,
+			}
+			topology.AddLink(p.Ctx.Graph, p.Groups[blockdevGroupName], node, "connected", linkMetadata)
+			if !topology.HaveLayer2Link(p.Ctx.Graph, p.Groups[blockdevGroupName], node) {
+				linkMetadata := graph.Metadata{
+					"Type":      "blockdev",
+					"Bandwidth": 120,
+				}
+				topology.AddLayer2Link(p.Ctx.Graph, p.Groups[blockdevGroupName], node, linkMetadata)
+			}
+		}
+
+	}
+
+	for i := range blockdev.Children {
+		childNode := p.Ctx.Graph.LookupFirstNode(graph.Metadata{"Name": blockdev.Children[i].getName()})
+		if childNode != nil {
+			topology.AddLink(p.Ctx.Graph, childNode, node, "connected", nil)
+		}
+	}
+
+	p.blockdevMap[blockdev.Name] = blockdevInfo{
+		ID:   blockdev.Name,
+		Node: node,
+	}
+	return groupNode
+}
+
+// Prior to version 2.33 lsblk generated JSON that encodes all values
+// as strings.  Version 2.33 changed how integers and booleans are
+// encoded.  unmarshalWeakTypeJSON() will decode either version of the
+// JSON.
+func (p *ProbeHandler) unmarshalWeakTypeJSON(jsonBytes []byte) ([]BlockDevice, error) {
+	var decoder *mapstructure.Decoder
+	var err error
+	var result []BlockDevice
+
+	deviceMap := make(map[string]interface{})
+	err = json.Unmarshal([]byte(jsonBytes), &deviceMap)
+	config := &mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           &result,
+	}
+	if decoder, err = mapstructure.NewDecoder(config); err != nil {
+		return nil, err
+	}
+	if err := decoder.Decode(deviceMap["blockdevices"]); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (p *ProbeHandler) unregisterBlockdev(id string) {
+	p.Lock()
+	defer p.Unlock()
+
+	info, ok := p.blockdevMap[id]
+	if !ok {
+		return
+	}
+
+	p.Ctx.Graph.Lock()
+	if err := p.Ctx.Graph.DelNode(info.Node); err != nil {
+		p.Ctx.Graph.Unlock()
+		p.Ctx.Logger.Error(err)
+		return
+	}
+	p.Ctx.Graph.Unlock()
+
+	delete(p.blockdevMap, id)
+}
+
+func (p *ProbeHandler) connect() error {
+	var (
+		cmdOut []byte
+		err    error
+		intf   interface{}
+		result Devices
+	)
+
+	lsblkPath := p.Ctx.Config.GetString("agent.topology.blockdev.lsblk_path")
```

Comment:
I think you should check that this is not empty, I think it is possible to set an empty value in the config

Reply:
Ok.